File: 150.evaluate-reverse-polish-notation.js
/*
 * @lc app=leetcode id=150 lang=javascript
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
    const stack = [];

    for (let i = 0; i < tokens.length; i++) {
        if (isNaN(tokens[i])) {
            const num2 = Number(stack.pop());
            const num1 = Number(stack.pop());
            const res = calc(num1, num2, tokens[i]);
            stack.push(res);
        } else {
            stack.push(tokens[i])
        }
    };

    return stack[0]

};
// @lc code=end

/**
 * @param {number} num1
 * @param {number} num2
 * @param {string} operator
 * @return {number}
 */
const calc = (num1, num2, operator) => { 
    if (operator === "+") return num1 + num2;
    if (operator === "-") return num1 - num2; 
    if (operator === "*") return num1 * num2;
    if (operator === "/") return Math.trunc(num1 / num2);
}

File: python/150.evaluate-reverse-polish-notation/150.evaluate-reverse-polish-notation.py
#
# @lc app=leetcode id=150 lang=python3
#
# [150] Evaluate Reverse Polish Notation
#

# @lc code=start
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        addition = "+"
        subtraction = "-"
        multiplier = "*"
        dividion = "/"

        stack = []

        for token in tokens:
            
            if token not in [addition, subtraction, multiplier, dividion]:
                stack.append(int(token))
                continue
        
            if token == addition:
                sum = stack.pop() + stack.pop()
                stack.append(sum)

            elif token == subtraction:
                letter = stack.pop()
                reminder = stack.pop() - letter
                stack.append(reminder)

            elif token == multiplier:
                result = stack.pop() * stack.pop()
                stack.append(result)

            else:
                letter = stack.pop()
                result = stack.pop() / letter
                stack.append(math.floor(result) if result > 0 else math.floor(abs(result)) * -1)
                
        return stack.pop()
# @lc code=end
